pkg/agent/instancevalidator.go

```diff
@@ -55,39 +54,67 @@ func (j *instanceValidator) validateAPIOnDataplane() {
 		}
 		externalAPIID := serviceInstanceResource.Attributes[definitions.AttrExternalAPIID]
 		externalAPIStage := serviceInstanceResource.Attributes[definitions.AttrExternalAPIStage]
+		externalPrimaryKey := serviceInstanceResource.Attributes[definitions.AttrExternalAPIPrimaryKey]
 		// Check if the consumer instance was published by agent, i.e. following attributes are set
 		// - externalAPIID should not be empty
 		// - externalAPIStage could be empty for dataplanes that do not support it
 		if externalAPIID != "" && !agent.apiValidator(externalAPIID, externalAPIStage) {
-			j.deleteServiceInstanceOrService(serviceInstanceResource, externalAPIID)
+			j.deleteServiceInstanceOrService(serviceInstanceResource, externalPrimaryKey, externalAPIID, externalAPIStage)
 		}
 	}
 }
 
-func (j *instanceValidator) shouldDeleteService(apiID string) bool {
-	list, err := agent.apicClient.GetConsumerInstancesByExternalAPIID(apiID)
-	if err != nil {
-		return false
+func (j *instanceValidator) shouldDeleteService(externalAPIPrimaryKey, externalAPIID, externalAPIStage string) bool {
+	instanceCount := 0
+	if externalAPIPrimaryKey != "" {
+		instanceCount = j.getServiceInstanceCount(definitions.AttrExternalAPIPrimaryKey, externalAPIPrimaryKey)
+		log.Tracef("Query instances with externalPrimaryKey attribute : %s", externalAPIPrimaryKey)
+	} else {
+		instanceCount = j.getServiceInstanceCount(definitions.AttrExternalAPIID, externalAPIID)
+		log.Tracef("Query instances with externalAPIID attribute : %s", externalAPIID)
 	}
 
-	// if there is only 1 consumer instance left, we can signal to delete the service too
-	return len(list) <= 1
+	log.Tracef("Instances count : %d", instanceCount)
+
+	return instanceCount <= 1
+}
+
+func (j *instanceValidator) getServiceInstanceCount(attName, attValue string) int {
+	instanceCount := 0
+	for _, key := range agent.cacheManager.GetAPIServiceInstanceKeys() {
+		serviceInstanceResource, _ := agent.cacheManager.GetAPIServiceInstanceByID(key)
+		if serviceInstanceResource != nil {
+			instaceAttValue := serviceInstanceResource.Attributes[attName]
```

Comment:
Just a heads up. I moved attributes into a SubResource for the instances. You might want to sync up with my branch and update this line.